ModFrmHilD/PositiveUnitReps.m

```diff
@@ -0,0 +1,677 @@
+end intrinsic;
+
+intrinsic RepIdealConversion(M::ModFrmHilDGRng) -> Assoc, Assoc
+  {
+    inputs:
+      M: A graded ring of Hilbert modular forms
+    returns:
+      Two 2D associative arrays,
+        - IdealToRep
+        - RepToIdeal,
+      which cache the conversion of nn to nu. 
+      Precisely, for each nn in the ideal class 
+      [bbp]^-1, the ideal nn * bbp is narrowly
+      principal, and we can use FunDomainRep
+      to get a totally positive generator.
+      We have IdealToRep[bb][nu] = nn
+      and RepToIdeal[bb][nu] = nn
+
+      // TODO we could combine this function into
+      // FunDomainRepsUpToNorm and maybe save a loop
+      // or two, but because narrow principalizing
+      // is probably the most expensive step anyways
+      // I figured code readability/cleanliness was
+      // worth more. There is a known optimization here
+      // if this step is too slow though. 
+  }
+  
+  if assigned M`IdealToRep and M`RepToIdeal then
+    return M`RepToIdeal, M`IdealToRep;
+  end if;
+
+  M`IdealToRep := AssociativeArray();
+  M`RepToIdeal := AssociativeArray();
+
+  F := BaseField(M);
+  ZF := Integers(M);
+  dd := Different(ZF);
+
+  for bb in NarrowClassGroupReps(M) do
+    M`IdealToRep[bb] := AssociativeArray();
+    M`RepToIdeal[bb] := AssociativeArray();
+
+    // dealing with the zero ideal, which lives
+    // on every component
+    M`IdealToRep[bb][0*ZF] := F!0;
+    M`RepToIdeal[bb][F!0] := 0*ZF;
+  end for;
+
+  for nn in IdealsUpTo(M`Precision, ZF) do
+    // we've already dealt with the zero ideal
+    if IsZero(nn) then
+      continue;
+    end if;
+    bb := IdealToNarrowClassRep(M, nn);
+    bbp := bb * dd^-1;
+    _, nu := IsNarrowlyPrincipal(nn * bbp);
+    nu, _ := FunDomainRep(nu);
+    M`IdealToRep[bb][nn] := nu;
+    M`RepToIdeal[bb][nu] := nn;
+  end for;
+
+  return M`RepToIdeal, M`IdealToRep;
+end intrinsic;
+
+intrinsic FunDomainReps(M::ModFrmHilDGRng) -> Assoc
+  {getter}
+  return M`FunDomainReps;
+end intrinsic;
+
+intrinsic FunDomainRepsOfNorm(M::ModFrmHilDGRng, bb::RngOrdFracIdl, x::RngIntElt) -> SeqEnum
+  {
+    inputs:
+      M: A graded ring of Hilbert modular forms
+      bb: An ideal class representative of M
+      x: An integer norm
+    returns:
+      The fundamental domain representatives nu such that the integral ideal
+      nn associated to nu has norm x.
+  }
+  return FunDomainRepsOfNorm(M)[bb][x];
+end intrinsic;
+
+intrinsic FunDomainRepsUpToNorm(M::ModFrmHilDGRng, bb::RngOrdFracIdl, x::RngIntElt) -> SeqEnum
+  {
+    inputs:
+      M: A graded ring of Hilbert modular forms
+      bb: An ideal class representative of M
+      x: An integer norm
+    returns:
+      The fundamental domain representatives nu such that the integral ideal
+      nn associated to nu has norm at most x.
+  }
+  return FunDomainRepsUpToNorm(M)[bb][x];
+end intrinsic;
+
+intrinsic FunDomainRepsUpToNorm(M::ModFrmHilDGRng, bb::RngOrdFracIdl, x::RngIntElt) -> SetEnum
```

Comment:
The initial version of the code only stored `FunDomainRepsUpToNorm`,  at a `SetEnum` and computed `FunDomainRepsOfNorm` by computing the difference (which is fast for `SetEnum`s). However, there are places in the code (e.g. Hecke operators) where we want the representatives ordered by norm, and I think that's hard to enforce from a `SetEnum`.  As such, I changed it to a `SeqEnum`, but then this slows down the computation of reps of a given norm, which gets used in the shadows computation. Therefore, I concluded that it'd be best to just store both, but I agree that it's not great and would love a better solution.

Reply:
(removing duplicate)

Reply:
I think it's fine. I was just a bit surprised when reading this at first.
Would an `AssociativeArray` with norms as keys work?

Reply:
You mean having `FunDomainRepsOfNorm` be an associative array with norms as keys and objects `SetEnum`s and then joining then whenever we need reps up to? 

I think this would work and I'm okay with it. I decided against this because `FunDomainRepsUpToNorm` gets called a lot, whereas `FunDomainRepsOfNorm` is only really used during the precomputation I think.